### ISSUE TITLE
Borrow multiple Rust fields (on different objects) on the same `env` concurrently?

### ISSUE THREAD

**Issue by MarijnS95** [jni-rs/jni-rs#539]
I just ran into a "cannot borrow `env` as mutable more than once at a time" error with the following code (simplified for brevity):

```rust
#[no_mangle]
pub extern "system" fn Java_...render(
    mut env: JNIEnv,
    _class: JClass,
    renderer: JObject,
    surface: JObject,
) {
    let mut renderer = unsafe { env.get_rust_field(renderer, "mNative") }.unwrap();
    let surface = unsafe { env.get_rust_field(surface, "mNative") }.unwrap();
    renderer.render(&surface)
}
```

The second call to `get_rust_field()` fails because `env` is already mutably borrowed for the first, but I'd like to have both locks in scope at the same time to call a function that uses both Rust objects. Is this intended, and is there a way around it? So far I've found `let mut env2 = unsafe { env.unsafe_clone() };` to work and "be safe" as I'm not "creating any local references" as far as I understand?

**Comment by argv-minus-one** [comment]
Yes. `get_rust_field` doesn't create any local references, so using `unsafe_clone` is appropriate.

**Comment by rib** [comment]
Just looking at how this would look with the changes for 0.22, it's maybe worth noting that `get_rust_field` no longer requires a `&mut Env` reference since these APIs don't need to return a new local reference (i.e. they don't need to modify the top JNI stack frame) (ref https://github.com/jni-rs/jni-rs/issues/455).

The following code is a rough adaptation of the example, including using `EnvUnowned::with_env` to wrap the native method in a catch_unwind:

```rust
use jni::{objects::JObject, EnvUnowned};

pub struct Renderer {
    // fields
}
impl Renderer {
    pub fn render(&mut self, _surface: &Surface) {
        // rendering logic
    }
}

pub struct Surface {
    // fields
}

#[no_mangle]
pub extern "system" fn Java_Renderer_render<'local>(
    mut unowned_env: EnvUnowned<'local>,
    _this: JObject<'local>,
    renderer: JObject<'local>,
    surface: JObject<'local>,
) {
    unowned_env
        .with_env(|env| -> jni::errors::Result<_> {
            let mut renderer = unsafe { env.get_rust_field::<_, _, Renderer>(renderer, c"mNative")? };
            let surface = unsafe { env.get_rust_field::<_, _, Surface>(surface, c"mNative")? };
            renderer.render(&surface);
            Ok(())
        })
        .resolve::<jni::errors::LogErrorAndDefault>()
}
```
(So essentially the code in question from the example should work, ignoring the unrelated `EnvUnowned` changes)

It could still be annoying if you lock several fields and then also want to use another API that takes a `&mut Env` like `env.new_string(c"foo")?` but you could potentially then use `env.with_local_frame(8, |env| {})` or `env.with_top_local_frame(|env|{})` to get a new scope with a mutable Env without `unsafe` code.

I think this can be improved further though because there's no need for the returned `MutexGuard<'env>` to borrow from the `&'env Env`.

For all intents and purposes the `MutexGuard` we will get internally will have a `'static` lifetime (we've handed a Boxed allocation containing a `Mutex` to the JVM).

We could let the `get_rust_field` caller name the lifetime with anything they like but probably a better choice would be the local reference frame lifetime from the `Env<'local>` so you would at least be expected to unlock the field before unwinding the current JNI stack frame. Using that `local` lifetime would avoid borrowing the `Env` itself though, so if you have a `&mut Env` then you will still have a `&mut Env` after `get_rust_field` returns.
